<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <meta name="author" content="Timo Lassmann">
  <title>RNAseq Pipeline</title>
  <style type="text/css">code{white-space: pre;}</style>
  <!--[if lt IE 9]>
    <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  <style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
  </style>
  <link rel="stylesheet" href="../../reproducibility/doc/style.css">
</head>
<body>
<header>
<h1 class="title">RNAseq Pipeline</h1>
<h2 class="author">Timo Lassmann</h2>
<h3 class="date">Dec. 5th 2013</h3>
</header>
<nav id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#some-details">Some details</a></li>
<li><a href="#installation-and-usage">Installation and Usage</a></li>
<li><a href="#results-and-code">Results and CODE</a><ul>
<li><a href="#pipeline-paramenters-and-settings">Pipeline Paramenters and Settings</a><ul>
<li><a href="#cufflinks-command">Cufflinks Command</a></li>
<li><a href="#cuffquant-command">Cuffquant Command</a></li>
<li><a href="#paths">Paths</a></li>
<li><a href="#summarydirectory">SummaryDirectory</a></li>
<li><a href="#files">Files</a></li>
</ul></li>
<li><a href="#postanalysis">Postanalysis</a></li>
</ul></li>
</ul>
</nav>
<h1 id="introduction">Introduction</h1>
<p>Analysis of next generation sequencing datasets is complicated and often hard to reproduce. This package contains a pipeline system to process RNAseq data in a <em>totally</em> reproducible manner. Moreover, the package contains all the code, scripts and parameters to run the analysis. Typing <strong>make</strong> in the root directory will actually start the entire analysis automatically.</p>
<p>For convenience during development all steps are executed within <strong>makefiles</strong> ensuring that computationally expensive initial steps do not have to be repeated if downstream analysis steps are modified. The makefiles also keep the documentation up to date.</p>
<p>Finally I use the literal programming paradigm to describe and highlight the important steps. This file actually contains the code needed for the analysis.</p>
<h1 id="some-details">Some details</h1>
<p>The package is organized as follows:</p>
<ol type="1">
<li>the configure.ac file contains the instructions to check the versions of external programs used in the pipeline. As a default if the installed version is higher than the required version a warning will be produced. Otherwise the configuration will fail.</li>
<li>once configured, make will compile the C code used to extract code from this document.</li>
<li>After all programs are extracted ro the <strong>src</strong> directory, a makefile will call the <strong>run.mk</strong> makefile and start the pipelines.</li>
</ol>
<h1 id="installation-and-usage">Installation and Usage</h1>
<p>To install the package simple follow the steps below:</p>
<pre class="sourceCode bash"><code class="sourceCode bash">
<span class="kw">tar</span> -zxvf rnaseq_pipeline-XX.XX.tgz
<span class="kw">cd</span> rnaseq_pipeline-XX.XX
<span class="kw">./configure</span></code></pre>
<p><strong>WARNING</strong>: Typing <strong>make</strong> now will not only compile or extract code but will actually start running pipelines on the input directories specifies below</p>
<pre class="sourceCode bash"><code class="sourceCode bash">
<span class="kw">make</span></code></pre>
<p>To apply the same pipeline to different data or change parameter settings simply edit the corresponding sections in <a href="#pipeline-paramenters-and-settings">this file</a>.</p>
<h1 id="results-and-code">Results and CODE</h1>
<p>The pipeline consists of one main makefile called <strong>run.mk</strong>. Within this files larger analysis units are individually executed using secondary <strong>makefiles</strong>. Output files from one step form the dependency of the next step.</p>
<h2 id="pipeline-paramenters-and-settings">Pipeline Paramenters and Settings</h2>
<h4 id="input-data">Input Data</h4>
<p>We will run the pipeline on the following two MiSeq runs. Later we repeat with the 150bp paired end HiSeq rapid runs.</p>
<h4 id="tophat-command">Tophat Command</h4>
<p>The Tophat command used is:</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
    <span class="ch">$(</span><span class="dt">TOPHAT</span><span class="ch">)</span> -p 1 --b2-very-sensitive  <span class="dt">--output-dir</span><span class="ch">=$(</span><span class="dt">OUTDIR</span><span class="ch">)</span><span class="st">/</span><span class="ch">$@</span><span class="st">  --transcriptome-index=</span><span class="ch">$(</span><span class="dt">GENCODE_DIR</span><span class="ch">)</span><span class="st"> </span><span class="ch">$(</span><span class="dt">GENOME</span><span class="ch">)</span><span class="st"> </span><span class="ch">$^</span></code></pre>
<h3 id="cufflinks-command">Cufflinks Command</h3>
<p>The actual Cufflinks command used.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
    <span class="ch">$(</span><span class="dt">CUFFLINKS</span><span class="ch">)</span> -o <span class="ch">$(</span><span class="dt">OUTDIR</span><span class="ch">)</span>/<span class="ch">$@</span>   -g <span class="ch">$(</span><span class="dt">GENCODE_GTF</span><span class="ch">)</span>  <span class="ch">$^</span>/accepted_hits.bam
    </code></pre>
<h3 id="cuffquant-command">Cuffquant Command</h3>
<p>The actual Cuffquant command used.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
    <span class="ch">$(</span><span class="dt">CUFFQUANT</span><span class="ch">)</span> -o <span class="ch">$(</span><span class="dt">OUTDIR</span><span class="ch">)</span>/<span class="ch">$@</span>   <span class="ch">$(</span><span class="dt">GENCODE_GTF</span><span class="ch">)</span>  <span class="ch">$^</span>/accepted_hits.bam
    </code></pre>
<h3 id="paths">Paths</h3>
<p>Set the variables below according to your installation.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="dt">TOPHAT</span><span class="ch">=</span><span class="st">/usr/local/bin/tophat</span>
<span class="dt">CUFFLINKS</span><span class="ch">=</span><span class="st">/usr/local/bin/cufflinks</span>
<span class="dt">CUFFQUANT</span><span class="ch">=</span><span class="st">/usr/local/bin/cuffquant</span>

<span class="dt">GENCODE_GTF</span><span class="ch">=</span><span class="st">../src/customgencode.gtf</span>
<span class="dt">GENCODE_DIR</span><span class="ch">=</span><span class="st">../src/customgencode</span>

<span class="dt">GENOME</span><span class="ch">=</span><span class="st">../src/genome</span>
</code></pre>
<h3 id="summarydirectory">SummaryDirectory</h3>
<p>This directory will collect all summary files across multiple runs. This code snippet is used in mulitple post-processing scripts.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="dt">SUMMARYDIR</span><span class="ch">=</span><span class="st">../summary/</span></code></pre>
<h3 id="files">Files</h3>
<p>Here is the code of the actual scripts used in the analysis.</p>
<h4 id="file-run.mk">File: run.mk</h4>
<p>The makefile <strong>run.mk</strong> executes the pipeline <strong>process_run.mk</strong> on all target libraries. Afterwards additional makefiles collect output files and summarize the results.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">


<span class="ot">.PHONY:</span><span class="dt"> mapping stats genefpkm</span>

<span class="dv">all:</span><span class="dt"> mapping stats normtable</span>

<span class="dv">mapping:</span>
    <span class="ch">$(</span><span class="dt">MAKE</span><span class="ch">)</span> -f makegenome.mk
    <span class="ch">$(</span><span class="dt">MAKE</span><span class="ch">)</span> -f mirrordata.mk
    <span class="ch">$(</span><span class="dt">MAKE</span><span class="ch">)</span> -k -j 80 indir=../data/  -f process_run.mk
    
<span class="dv">normtable:</span>
    <span class="ch">$(</span><span class="dt">MAKE</span><span class="ch">)</span> -f make_norm_table.mk

<span class="dv">stats:</span><span class="dt"> </span>
    <span class="ch">$(</span><span class="dt">MAKE</span><span class="ch">)</span> -f get_mapping_stats.mk

<span class="dv">genefpkm:</span>
    <span class="ch">$(</span><span class="dt">MAKE</span><span class="ch">)</span> -f join_gene_fpkm.mk
</code></pre>
<h4 id="file-process_run.mk">File: process_run.mk</h4>
<p>The makefile to process the output of one HiSeq or MiSeq run. The data is assumed to be already de-multiplexed.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="co">## Paths</span>

<span class="ot">.PHONY:</span><span class="dt"> directories alignment_stats</span></code></pre>
<p>Check of input parameteres are set.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">

<span class="kw">ifndef</span> indir
<span class="ch">$(</span><span class="kw">error</span><span class="st"> indir is not set</span><span class="ch">)</span>
<span class="kw">endif</span>

<span class="dt">OUTDIR</span><span class="ch">=</span><span class="st">../mapping/</span>

<span class="dt">dirs </span><span class="ch">:=</span><span class="st"> </span><span class="ch">$(</span><span class="dt">indir</span><span class="ch">)</span>

<span class="dt">files </span><span class="ch">:=</span><span class="st"> </span><span class="ch">$(</span><span class="kw">foreach</span><span class="st"> dir</span><span class="kw">,</span><span class="ch">$(</span><span class="dt">dirs</span><span class="ch">)</span><span class="kw">,</span><span class="st">      </span><span class="ch">$(</span><span class="kw">sort</span><span class="st"> </span><span class="ch">$(</span><span class="kw">wildcard</span><span class="st"> </span><span class="ch">$(</span><span class="kw">dir</span><span class="ch">)</span><span class="st">/*fq</span><span class="ch">)))</span>

</code></pre>
<p>Set targets for TopHat and Cufflinks.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="dt">TOPHAT_OUT</span><span class="ch">=$(</span><span class="kw">notdir</span><span class="st"> </span><span class="ch">$(</span><span class="kw">patsubst</span><span class="st"> %.fq</span><span class="kw">,</span><span class="st">%_tophat_out</span><span class="kw">,</span><span class="st">  </span><span class="ch">$(</span><span class="dt">files</span><span class="ch">)))</span>


<span class="dt">CUFFQUANT_OUT</span><span class="ch">=$(</span><span class="kw">notdir</span><span class="st"> </span><span class="ch">$(</span><span class="kw">patsubst</span><span class="st"> %.fq</span><span class="kw">,</span><span class="st">%_cuffquant_out</span><span class="kw">,</span><span class="st">  </span><span class="ch">$(</span><span class="dt">files</span><span class="ch">)))</span></code></pre>
<p>Add directories to <strong>vpath</strong> to allow for targets and prerequisites to be in different directories.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
vpath %.fq  <span class="ch">$(</span><span class="dt">indir</span><span class="ch">)</span>
vpath %.fq  <span class="ch">$(</span><span class="dt">indir</span><span class="ch">)</span>
vpath %_tophat_out  <span class="ch">$(</span><span class="dt">OUTDIR</span><span class="ch">)</span>

vpath %_cuffquant_out  <span class="ch">$(</span><span class="dt">OUTDIR</span><span class="ch">)</span>

<span class="dv">all:</span><span class="dt"> directories </span><span class="ch">$(</span><span class="dt">TOPHAT_OUT</span><span class="ch">)</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">CUFFQUANT_OUT</span><span class="ch">)</span>
    <span class="ch">@</span><span class="fu">echo </span><span class="ch">$(</span><span class="dt">OUTDIR</span><span class="ch">)</span>

<span class="dv">%_tophat_out:</span><span class="dt"> %.fq</span>
<span class="co">## Tophat Command</span>



<span class="dv">%_cuffquant_out:</span><span class="dt"> %_tophat_out</span>
<span class="co">## Cuffquant Command</span>


<span class="dt">MKDIR_P </span><span class="ch">=</span><span class="st"> mkdir -p</span>

<span class="dv">directories:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">OUTDIR</span><span class="ch">)</span>

<span class="dv">$(OUTDIR):</span>
    <span class="ch">$(</span><span class="dt">MKDIR_P</span><span class="ch">)</span> <span class="ch">$(</span><span class="dt">OUTDIR</span><span class="ch">)</span></code></pre>
<h4 id="file-join_gene_fpkm.mk">File: join_gene_fpkm.mk</h4>
<p>Makefile to create the basic gene based fpkm expression table. For some reason adding awk code directly into this makefile did not work. Therefore I call the awk script <strong>join_gene_fpkm.awk</strong> from this file.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="co">## SummaryDirectory</span>

<span class="dt">GENEFPKM</span><span class="ch">=$(</span><span class="kw">shell</span><span class="st"> find ../mapping/ -name &#39;genes.fpkm_tracking&#39;</span><span class="ch">)</span>

<span class="kw">ifeq</span> <span class="st">&quot;</span><span class="ch">$(</span><span class="dt">GENEFPKM</span><span class="ch">)</span><span class="st">&quot;</span> <span class="st">&quot;&quot;</span>
<span class="dv">all:</span>
    <span class="ch">@</span><span class="fu">echo </span><span class="st">&quot;Warning No FPKM files found.&quot;</span>
<span class="kw">else</span>
<span class="dv">all:</span><span class="dt">directories </span><span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span><span class="dt">/all_gene.fpkm_tracking</span>

<span class="dv">$(SUMMARYDIR)/all_gene.fpkm_tracking:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">GENEFPKM</span><span class="ch">)</span>
    awk -f   join_gene_fpkm.awk <span class="ch">$^</span> &gt; <span class="ch">$@</span>
<span class="kw">endif</span>

<span class="dt">MKDIR_P </span><span class="ch">=</span><span class="st"> mkdir -p</span>

<span class="dv">directories:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span>

<span class="dv">$(SUMMARYDIR):</span>
    <span class="ch">$(</span><span class="dt">MKDIR_P</span><span class="ch">)</span> <span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span></code></pre>
<h4 id="file-get_mapping_stats.mk">File: get_mapping_stats.mk</h4>
<p>Collects the mapping statistics from all mapped libraries.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="co">## SummaryDirectory</span>

<span class="dt">ALIGNMENTSTATS</span><span class="ch">=$(</span><span class="kw">shell</span><span class="st"> find ../mapping/ -name &#39;align_summary.txt&#39;</span><span class="ch">)</span>

<span class="kw">ifeq</span> <span class="st">&quot;</span><span class="ch">$(</span><span class="dt">ALIGNMENTSTATS</span><span class="ch">)</span><span class="st">&quot;</span> <span class="st">&quot;&quot;</span>
<span class="dv">all:</span>
    <span class="ch">@</span><span class="fu">echo </span><span class="st">&quot;Warning No Mapping Stats&quot;</span>
<span class="kw">else</span>
<span class="dv">all:</span><span class="dt">directories </span><span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span><span class="dt">/all_mapping_stats.csv </span><span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span><span class="dt">/all_mapping_stats_counts.csv</span>
    <span class="ch">@</span><span class="fu">echo </span><span class="st">&quot;Done.&quot;</span>

<span class="dv">$(SUMMARYDIR)/all_mapping_stats.csv:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">ALIGNMENTSTATS</span><span class="ch">)</span>
    awk <span class="st">&#39;{if(</span><span class="ch">$$</span><span class="st">2 == &quot;concordant&quot; &amp;&amp; </span><span class="ch">$$</span><span class="st">3 == &quot;pair&quot;){sub(&quot;_tophat_out/align_summary.txt&quot;,&quot;&quot;,FILENAME); printf &quot;%s\t%s\n&quot; ,FILENAME,</span><span class="ch">$$</span><span class="st">1    }}&#39;</span> <span class="ch">$(</span><span class="dt">ALIGNMENTSTATS</span><span class="ch">)</span> &gt;<span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span>/all_mapping_stats.csv

<span class="dv">$(SUMMARYDIR)/all_mapping_stats_counts.csv:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">ALIGNMENTSTATS</span><span class="ch">)</span>
    awk <span class="st">&#39;{if(</span><span class="ch">$$</span><span class="st">1 == &quot;Aligned&quot; &amp;&amp; </span><span class="ch">$$</span><span class="st">2 == &quot;pairs:&quot;){sub(&quot;_tophat_out/align_summary.txt&quot;,&quot;&quot;,FILENAME); printf &quot;%s\t%s\n&quot; ,FILENAME,</span><span class="ch">$$</span><span class="st">3 }}&#39;</span> <span class="ch">$(</span><span class="dt">ALIGNMENTSTATS</span><span class="ch">)</span> &gt;<span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span>/all_mapping_stats_counts.csv.csv



<span class="kw">endif</span>

<span class="dt">MKDIR_P </span><span class="ch">=</span><span class="st"> mkdir -p</span>

<span class="dv">directories:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span>

<span class="dv">$(SUMMARYDIR):</span>
    <span class="ch">$(</span><span class="dt">MKDIR_P</span><span class="ch">)</span> <span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span></code></pre>
<h4 id="file-mirrordata.mk">File: mirrordata.mk</h4>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="dt">DATADIR</span><span class="ch">=</span><span class="st">../data</span>


<span class="dt">FILES</span><span class="ch">=$(</span><span class="kw">shell</span><span class="st"> find /data/lassmann/benchmark/STRT/extracted  -type f -name &#39;*.fq&#39; -size +1M</span><span class="ch">)</span>


<span class="dt">FASTQDIRS </span><span class="ch">=</span><span class="st"> </span><span class="ch">$(</span><span class="kw">dir</span><span class="st"> </span><span class="ch">$(</span><span class="dt">FILES</span><span class="ch">))</span>

<span class="dt">FASTQFILES </span><span class="ch">=</span><span class="st"> </span><span class="ch">$(</span><span class="kw">notdir</span><span class="st"> </span><span class="ch">$(</span><span class="dt">FILES</span><span class="ch">))</span>

<span class="dt">TARGETS </span><span class="ch">=</span><span class="st"> </span><span class="ch">$(</span><span class="kw">addprefix</span><span class="st"> </span><span class="ch">$(</span><span class="dt">DATADIR</span><span class="ch">)</span><span class="st">/</span><span class="kw">,</span><span class="st"> </span><span class="ch">$(</span><span class="dt">FASTQFILES</span><span class="ch">))</span>



<span class="kw">ifeq</span> <span class="st">&quot;</span><span class="ch">$(</span><span class="kw">words</span><span class="st">  </span><span class="ch">$(</span><span class="dt">FILES</span><span class="ch">))</span><span class="st">&quot;</span> <span class="st">&quot;0&quot;</span>
<span class="dv">all:</span>
    <span class="ch">@</span><span class="fu">echo </span><span class="st">&quot;Warning No Fastq files found.&quot;</span>
<span class="kw">else</span>
<span class="dv">all:</span><span class="dt"> directories </span><span class="ch">$(</span><span class="dt">TARGETS</span><span class="ch">)</span><span class="dt"> </span>
<span class="kw">endif</span>

    
<span class="dv">$(DATADIR)/%.fq:</span><span class="dt">   %.fq</span>
    cp -av <span class="ch">$&lt;</span> <span class="ch">$@</span>

<span class="dt">MKDIR_P </span><span class="ch">=</span><span class="st"> mkdir -p</span>

<span class="dv">directories:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">DATADIR</span><span class="ch">)</span>

<span class="dv">$(DATADIR):</span><span class="dt"> </span>
    <span class="ch">$(</span><span class="dt">MKDIR_P</span><span class="ch">)</span> <span class="ch">$(</span><span class="dt">DATADIR</span><span class="ch">)</span>

vpath %.fq <span class="ch">$(</span><span class="dt">FASTQDIRS</span><span class="ch">)</span></code></pre>
<h4 id="file-makegenome.mk">File: makegenome.mk</h4>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">


<span class="dt">TEST</span><span class="ch">=$(</span><span class="kw">shell</span><span class="st"> find . -name &#39;getgenomeucsc.sh&#39;</span><span class="ch">)</span>

<span class="kw">ifeq</span> <span class="st">&quot;</span><span class="ch">$(</span><span class="dt">TEST</span><span class="ch">)</span><span class="st">&quot;</span> <span class="st">&quot;&quot;</span>
<span class="dv">all:</span>
    <span class="ch">@</span><span class="fu">echo </span><span class="st">&quot;Warning No files for de-multiplexing found...&quot;</span>
<span class="kw">else</span>
<span class="dv">all:</span><span class="dt">  genome.1.bt2 customgencode.1.bt2</span>


<span class="kw">endif</span>
                                
<span class="dv">GRCm38.p2.genome.fa:</span>
    wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M2/GRCm38.p2.genome.fa.gz
    gunzip GRCm38.p2.genome.fa.gz

<span class="dv">gencode.vM2.annotation.gtf:</span>
    wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M2/gencode.vM2.annotation.gtf.gz
    gunzip gencode.vM2.annotation.gtf.gz

<span class="dv">genome.fa:</span><span class="dt"> GRCm38.p2.genome.fa</span>
    ./makecustomgenome GRCm38.p2.genome.fa -nogtf GRCm38.p2.genome.fa -o  genome

<span class="dv">customgencode.gtf:</span><span class="dt"> genome.fa  gencode.vM2.annotation.gtf</span>
    cat gencode.vM2.annotation.gtf  | awk <span class="st">&#39;{if(</span><span class="ch">$$</span><span class="st">26~/1|2/){  if(</span><span class="ch">$$</span><span class="st">3 !=&quot;gene&quot;){ print </span><span class="ch">$$</span><span class="st">0}}}&#39;</span>  &gt; tmp2.gtf
    grep <span class="st">&#39;&gt;&#39;</span> genome.fa | awk <span class="st">&#39;{x = substr(</span><span class="ch">$$</span><span class="st">0,2) ;print x}&#39;</span> &gt; chromosomes.txt
    grep -f  chromosomes.txt tmp2.gtf &gt; customgencode.gtf

<span class="dv">genome.1.bt2:</span><span class="dt"> customgencode.gtf genome.fa</span>
    bowtie2-build genome.fa genome
                                
<span class="dv">customgencode.1.bt2:</span><span class="dt"> customgencode.gtf</span>
    tophat -G customgencode.gtf --transcriptome-index=.  genome


                                </code></pre>
<h4 id="file-make_norm_table.mk">File: make_norm_table.mk</h4>
<p>Collects the mapping statistics from all mapped libraries.</p>
<pre class="sourceCode Makefile"><code class="sourceCode makefile">
<span class="ot">.PHONY:</span><span class="dt">cuffnorm</span>

<span class="co">## SummaryDirectory</span>

<span class="dt">CUFFNORM</span><span class="ch">=</span><span class="st">/usr/local/bin/cuffnorm</span>

<span class="dt">GENCODE_GTF</span><span class="ch">=</span><span class="st">../src/customgencode.gtf</span>
<span class="dt">QUANTDIR</span><span class="ch">=$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span><span class="st">/quantknowntable</span>

<span class="dt">INFILES</span><span class="ch">=$(</span><span class="kw">shell</span><span class="st"> find . -name &#39;abundances.cxb&#39; | awk &#39;BEGIN{print &quot;sample_name group&quot;} {split(</span><span class="ch">$$</span><span class="st">1,a,&quot;_cuffquant_out&quot;); printf &quot;%s\t%s\n&quot;,</span><span class="ch">$$</span><span class="st">1 ,substr(a[1],3,50 ) }&#39;</span><span class="ch">)</span>


<span class="kw">ifeq</span> <span class="st">&quot;</span><span class="ch">$(</span><span class="dt">INFILES</span><span class="ch">)</span><span class="st">&quot;</span> <span class="st">&quot;&quot;</span>
<span class="dv">all:</span>
    <span class="ch">@</span><span class="fu">echo </span><span class="st">&quot;Warning No Qumant found&quot;</span>
<span class="kw">else</span>
<span class="dv">all:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">QUANTDIR</span><span class="ch">)</span><span class="dt">/sample_sheet.txt cuffnorm</span>
    echo <span class="st">&quot;Done.&quot;</span>

<span class="kw">endif</span>

<span class="dv">cuffnorm:</span><span class="dt"> </span>
    <span class="ch">$(</span><span class="dt">CUFFNORM</span><span class="ch">)</span> -p 80 --use-sample-sheet -o <span class="ch">$(</span><span class="dt">QUANTDIR</span><span class="ch">)</span> <span class="ch">$(</span><span class="dt">GENCODE_GTF</span><span class="ch">)</span> <span class="ch">$(</span><span class="dt">QUANTDIR</span><span class="ch">)</span>/sample_sheet.txt

<span class="dv">$(QUANTDIR)/sample_sheet.txt:</span><span class="dt"> directories</span>
    find ../mapping/ -name <span class="st">&#39;abundances.cxb&#39;</span> | awk <span class="st">&#39;BEGIN{print &quot;sample_name group&quot;} {split(</span><span class="ch">$$</span><span class="st">1,a,&quot;_cuffquant_out&quot;); printf &quot;%s\t%s\n&quot; ,</span><span class="ch">$$</span><span class="st">1 ,substr(a[1],3,50 ) }&#39;</span> &gt; <span class="ch">$(</span><span class="dt">QUANTDIR</span><span class="ch">)</span>/sample_sheet.txt



<span class="dt">MKDIR_P </span><span class="ch">=</span><span class="st"> mkdir -p</span>

<span class="dv">directories:</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span><span class="dt"> </span><span class="ch">$(</span><span class="dt">QUANTDIR</span><span class="ch">)</span>

<span class="dv">$(QUANTDIR):</span>
    <span class="ch">$(</span><span class="dt">MKDIR_P</span><span class="ch">)</span>  <span class="ch">$(</span><span class="dt">QUANTDIR</span><span class="ch">)</span>

<span class="dv">$(SUMMARYDIR):</span>
    <span class="ch">$(</span><span class="dt">MKDIR_P</span><span class="ch">)</span> <span class="ch">$(</span><span class="dt">SUMMARYDIR</span><span class="ch">)</span>
</code></pre>
<h2 id="postanalysis">Postanalysis</h2>
<pre class="sourceCode R"><code class="sourceCode r">


<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(reshape)
<span class="kw">library</span>(gplots)
<span class="kw">library</span>(randomForest)
<span class="kw">library</span>(<span class="st">&quot;foreach&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;doSNOW&quot;</span>)
<span class="kw">registerDoSNOW</span>(<span class="kw">makeCluster</span>(<span class="dv">80</span>, <span class="dt">type=</span><span class="st">&quot;SOCK&quot;</span>))

<span class="kw">set.seed</span>(<span class="dv">42</span>)




fpkm_matrix &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;genes.fpkm_table&quot;</span>, <span class="dt">row.names=</span><span class="dv">1</span>)

dat =<span class="st"> </span><span class="kw">as.matrix</span>(fpkm_matrix[,<span class="dv">2</span>:<span class="kw">ncol</span>(fpkm_matrix)])
<span class="kw">class</span>(dat) &lt;-<span class="st"> &quot;numeric&quot;</span>
dat =<span class="st"> </span><span class="kw">t</span>(dat)


spikes =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(fpkm_matrix[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">2</span>:<span class="kw">ncol</span>(fpkm_matrix)]) )
spikes =<span class="st"> </span><span class="kw">cbind</span>(spikes, <span class="kw">substr</span>( <span class="kw">rownames</span>(spikes) ,<span class="dv">11</span>,<span class="dv">19</span>))

 
<span class="kw">colnames</span>(spikes) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Cherry&quot;</span>,<span class="st">&quot;Venus&quot;</span>,<span class="st">&quot;Spike2&quot;</span>, <span class="st">&quot;Spike6&quot;</span>,<span class="st">&quot;Spike4&quot;</span>,<span class="st">&quot;Spike5&quot;</span>,<span class="st">&quot;Spike7&quot;</span>,<span class="st">&quot;Spike1&quot;</span>, <span class="st">&quot;Spike3&quot;</span>,<span class="st">&quot;Spike8&quot;</span>,<span class="st">&quot;Run&quot;</span>)

<span class="kw">ggplot</span>(spikes, <span class="kw">aes</span>(Spike1+<span class="st"> </span><span class="fl">1e-05</span>, Spike4+<span class="st"> </span><span class="fl">1e-05</span>)) +<span class="st"> </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">facet_wrap</span>(<span class="st">&quot;Run&quot;</span>, <span class="dt">scale =</span> <span class="st">&quot;free&quot;</span>) +<span class="st"> </span><span class="kw">scale_x_log10</span>() +<span class="st"> </span><span class="kw">scale_y_log10</span>()

<span class="kw">ggsave</span>(<span class="st">&quot;Spike1_4.pdf&quot;</span>)

<span class="kw">ggplot</span>(spikes, <span class="kw">aes</span>(Cherry+<span class="st"> </span><span class="fl">1e-05</span>, Venus+<span class="st"> </span><span class="fl">1e-05</span>)) +<span class="st"> </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">facet_wrap</span>(<span class="st">&quot;Run&quot;</span>, <span class="dt">scale =</span> <span class="st">&quot;free&quot;</span>) +<span class="st"> </span><span class="kw">scale_x_log10</span>() +<span class="st"> </span><span class="kw">scale_y_log10</span>()
<span class="kw">ggsave</span>(<span class="st">&quot;Cherry_Venus.pdf&quot;</span>)


l =<span class="st"> </span><span class="kw">melt</span>(spikes)

<span class="kw">ggplot</span>(<span class="dt">data =</span> l, <span class="kw">aes</span>(<span class="dt">x =</span> variable, <span class="dt">y =</span> value +<span class="st"> </span><span class="fl">1e-05</span>)) +<span class="st"> </span><span class="kw">geom_boxplot</span>() +<span class="st"> </span><span class="kw">coord_flip</span>() +<span class="st"> </span><span class="kw">scale_y_log10</span>(<span class="st">&quot;Unit is fpkm&quot;</span>) +<span class="st"> </span><span class="kw">facet_wrap</span>(<span class="st">&quot;Run&quot;</span>) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;&quot;</span>)
<span class="kw">ggsave</span>(<span class="st">&quot;spikefpkm.pdf&quot;</span>)






<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(gplots)
<span class="kw">library</span>(randomForest)
<span class="kw">library</span>(<span class="st">&quot;foreach&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;doSNOW&quot;</span>)
<span class="kw">registerDoSNOW</span>(<span class="kw">makeCluster</span>(<span class="dv">80</span>, <span class="dt">type=</span><span class="st">&quot;SOCK&quot;</span>))

<span class="kw">set.seed</span>(<span class="dv">42</span>)


nzmean &lt;-<span class="st"> </span>function(x) {
    if (<span class="kw">all</span>(x==<span class="dv">0</span>)) <span class="dv">0</span> else <span class="kw">mean</span>(x)
}


nzcount &lt;-<span class="st"> </span>function(x) {
    <span class="kw">sum</span>(x !=<span class="dv">0</span>);
}

<span class="co">#mean = apply(x,1,nzmean)</span>

<span class="co">#nzcount &lt;- apply(x,1, nzcount)</span>


<span class="co">#either cell_cycle_genes_image.csv</span>

<span class="co">#mat = read.table(&quot;cell_cycle_genes_image.csv&quot;,header =T,row.names = 1,sep = &quot;\t&quot;)</span>

<span class="co">#or</span>

mat =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;all_gene.fpkm_tracking_and_Imaging.csv&quot;</span>,<span class="dt">header =</span>T,<span class="dt">row.names =</span> <span class="dv">1</span>,<span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
mat =<span class="st"> </span><span class="kw">t</span>(mat)

<span class="co">#end</span>

mat =<span class="st"> </span>mat[<span class="kw">rownames</span>(mat) !=<span class="st"> &quot;Description&quot;</span>,];

imagenamelist &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;SpotTotalAreaCh2&quot;</span>,<span class="st">&quot;SpotAvgAreaCh2&quot;</span>,<span class="st">&quot;SpotTotalIntenCh2&quot;</span>,<span class="st">&quot;SpotAvgIntenCh2&quot;</span>,<span class="st">&quot;SpotTotalAreaCh3&quot;</span>,<span class="st">&quot;SpotAvgAreaCh3&quot;</span>,<span class="st">&quot;SpotTotalIntenCh3&quot;</span>,<span class="st">&quot;SpotAvgIntenCh3&quot;</span>,<span class="st">&quot;ErrorType&quot;</span>,<span class="st">&quot;Description&quot;</span>,<span class="st">&quot;ImageKeep&quot;</span>,<span class="st">&quot;tracking_id.gene_short_name&quot;</span>);

rnadata =<span class="st"> </span>mat[!(<span class="kw">rownames</span>(mat) %in%<span class="st"> </span>imagenamelist),];


imagedata =<span class="st"> </span>mat[(<span class="kw">rownames</span>(mat) %in%<span class="st"> </span>imagenamelist),];



dat =<span class="st"> </span><span class="kw">as.matrix</span>(rnadata)
<span class="kw">class</span>(dat) &lt;-<span class="st"> &quot;numeric&quot;</span>
dat =<span class="st"> </span><span class="kw">t</span>(dat)


rf &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">ntree =</span> <span class="kw">rep</span>(<span class="dv">125</span>, <span class="dv">80</span>), <span class="dt">.combine =</span> combine, <span class="dt">.packages =</span> <span class="st">&quot;randomForest&quot;</span>) %dopar%<span class="st">  </span><span class="kw">randomForest</span>(dat,      <span class="dt">proximity =</span> <span class="ot">TRUE</span>,<span class="dt">ntree =</span> ntree)



fit =<span class="st"> </span><span class="kw">cmdscale</span>(<span class="dv">1</span> -<span class="st"> </span>rf$proximity)


kmeans =<span class="st"> </span><span class="kw">kmeans</span>(fit,<span class="dv">2</span>) 

<span class="kw">plot</span>(fit,<span class="dt">col =</span> kmeans$cluster)

smallest_cluster_size =<span class="st"> </span><span class="kw">min</span>(kmeans$size)


outlier =<span class="st"> </span>kmeans$cluster [kmeans$size[kmeans$cluster] ==smallest_cluster_size]


<span class="kw">names</span>(outlier);

rnadata_filtered =rnadata[,!(<span class="kw">colnames</span>(rnadata) %in%<span class="st"> </span><span class="kw">names</span>(outlier))]

<span class="co">#heatmap.2( log(dat+0.01),density.info=&quot;none&quot;,trace = c(&quot;none&quot;))</span>


imagedata_filtered =<span class="st"> </span>imagedata[,!(<span class="kw">colnames</span>(rnadata) %in%<span class="st"> </span><span class="kw">names</span>(outlier))]

good_image =<span class="st"> </span>imagedata_filtered[<span class="st">&quot;ImageKeep&quot;</span>,]  ==<span class="st"> </span><span class="dv">1</span>;


rnadata_filtered =<span class="st"> </span>rnadata_filtered[,good_image]
imagedata_filtered =<span class="st"> </span>imagedata_filtered[,good_image]
dat =<span class="st"> </span><span class="kw">as.matrix</span>(rnadata_filtered)
<span class="kw">class</span>(dat) &lt;-<span class="st"> &quot;numeric&quot;</span>



###extra filtering = remove genes with lower than 10 mean fpkm expression


dat_mean =<span class="st"> </span><span class="kw">apply</span>(dat,<span class="dv">1</span>,nzmean)


rnadata_filtered2 =<span class="st"> </span>rnadata_filtered[dat_mean &gt;<span class="st"> </span><span class="dv">10</span>,]

###


SpotAvgIntenCh2 =<span class="st"> </span>imagedata_filtered[<span class="st">&quot;SpotAvgIntenCh2&quot;</span>,]
SpotAvgIntenCh3 =<span class="st"> </span>imagedata_filtered[<span class="st">&quot;SpotAvgIntenCh3&quot;</span>,]

<span class="kw">class</span>(SpotAvgIntenCh2) &lt;-<span class="st"> &quot;numeric&quot;</span>

<span class="kw">class</span>(SpotAvgIntenCh3) &lt;-<span class="st"> &quot;numeric&quot;</span>

image_ratio =<span class="st"> </span><span class="kw">log2</span>((SpotAvgIntenCh2<span class="fl">+0.01</span>) /<span class="st"> </span>(SpotAvgIntenCh3<span class="fl">+0.01</span>))
<span class="kw">rownames</span>(image_ratio) =<span class="st"> &quot;image_ratio&quot;</span>
dat =<span class="st"> </span><span class="kw">as.matrix</span>(rnadata_filtered2)
<span class="kw">class</span>(dat) &lt;-<span class="st"> &quot;numeric&quot;</span>




dat=<span class="st"> </span><span class="kw">rbind</span>(dat,image_ratio)
frame =<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">t</span>(dat))
y &lt;-<span class="st"> </span>frame[, <span class="kw">ncol</span>(frame)]
x &lt;-<span class="st"> </span>frame[,<span class="dv">1</span>:(<span class="kw">ncol</span>(frame)-<span class="dv">1</span>)]
rf &lt;-<span class="st"> </span><span class="kw">randomForest</span>(x,y, <span class="dt">do.trace =</span> <span class="dv">100</span>,      <span class="dt">importance =</span> <span class="ot">TRUE</span>, <span class="dt">ntree =</span> <span class="dv">1</span>,<span class="dt">type=</span>regression)

<span class="co">#&gt; library(&quot;foreach&quot;)</span>
<span class="co">#&gt; library(&quot;doSNOW&quot;)</span>
<span class="co">#&gt; registerDoSNOW(makeCluster(80, type=&quot;SOCK&quot;))</span>

<span class="co">#&gt; x &lt;- matrix(runif(500), 100)</span>
<span class="co">#&gt; y &lt;- gl(2, 50)</span>

<span class="co">#&gt; rf &lt;- foreach(ntree = rep(250, 4), .combine = combine, .packages = &quot;randomForest&quot;) %dopar%</span>
<span class="co">#+    randomForest(x, y, ntree = ntree)</span>
<span class="co">#&gt; rf</span>
<span class="co">#Call:</span>
<span class="co">#randomForest(x = x, y = y, ntree = ntree)</span>
<span class="co">#Type of random forest: classification</span>


rf &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">ntree =</span> <span class="kw">rep</span>(<span class="dv">125</span>, <span class="dv">80</span>), <span class="dt">.combine =</span> combine, <span class="dt">.packages =</span> <span class="st">&quot;randomForest&quot;</span>) %dopar%<span class="st">    </span><span class="kw">randomForest</span>(x, y, <span class="dt">do.trace =</span> <span class="dv">100</span>,      <span class="dt">importance =</span> <span class="ot">TRUE</span>, <span class="dt">type=</span>regression,<span class="dt">ntree =</span> ntree)





<span class="kw">jpeg</span>(<span class="dt">filename=</span><span class="st">&quot;ALLRFaccuracy.jpg&quot;</span>, <span class="dt">width=</span><span class="dv">800</span>, <span class="dt">height=</span><span class="dv">800</span>, <span class="dt">pointsize=</span><span class="dv">12</span>,<span class="dt">bg=</span><span class="st">&quot;white&quot;</span>);
<span class="kw">plot</span>( <span class="kw">predict</span>(rf), y)
<span class="kw">abline</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>),<span class="dt">col=</span><span class="dv">2</span>)

<span class="kw">dev.off</span>();

<span class="kw">jpeg</span>(<span class="dt">filename=</span><span class="st">&quot;ALLRFimpvar.jpg&quot;</span>, <span class="dt">width=</span><span class="dv">800</span>, <span class="dt">height=</span><span class="dv">800</span>, <span class="dt">pointsize=</span><span class="dv">12</span>,<span class="dt">bg=</span><span class="st">&quot;white&quot;</span>);

<span class="kw">varImpPlot</span>(rf)
<span class="kw">dev.off</span>();

good_genes =<span class="st"> </span>rf$importance[,<span class="dv">1</span>]  &gt;<span class="st"> </span><span class="fl">0.1</span>

plotdata =<span class="st"> </span>rnadata_filtered2[good_genes,]

<span class="co">#plotdata = log(plotdata+0.1)</span>

plotdata =<span class="st"> </span><span class="kw">rbind</span>(plotdata,image_ratio)
gaga =<span class="st"> </span><span class="kw">as.matrix</span>(plotdata);

<span class="kw">class</span>(gaga) &lt;-<span class="st"> &quot;numeric&quot;</span>
<span class="kw">heatmap.2</span>( gaga,<span class="dt">density.info=</span><span class="st">&quot;none&quot;</span>,<span class="dt">trace =</span> <span class="kw">c</span>(<span class="st">&quot;none&quot;</span>))

test =<span class="st"> </span>( <span class="kw">cor</span>(<span class="kw">t</span>(gaga),<span class="dt">method=</span><span class="st">&quot;spearman&quot;</span>))


<span class="kw">jpeg</span>(<span class="dt">filename=</span><span class="st">&quot;ImportantGenesHeatmap.jpg&quot;</span>, <span class="dt">width=</span><span class="dv">1200</span>, <span class="dt">height=</span><span class="dv">800</span>, <span class="dt">pointsize=</span><span class="dv">12</span>,<span class="dt">bg=</span><span class="st">&quot;white&quot;</span>);
<span class="kw">heatmap.2</span>( test,<span class="dt">density.info=</span><span class="st">&quot;none&quot;</span>,<span class="dt">trace =</span> <span class="kw">c</span>(<span class="st">&quot;none&quot;</span>),<span class="dt">col=</span><span class="kw">redgreen</span>(<span class="dv">1000</span>))
<span class="kw">dev.off</span>()

## negative contriolll...




<span class="kw">sapply</span> (<span class="dv">1</span>:<span class="kw">ncol</span>(x), function (col) x[,col]&lt;&lt;-<span class="kw">sample</span>(x[,col]))




###ok

</code></pre>
</body>
</html>
